src/Voron/Data/Tables/Table.cs

```diff
@@ -1684,30 +1684,153 @@ public bool SeekOnePrimaryKeyWithPrefix(Slice prefix, Slice value, out TableValu
             var fst = GetFixedSizeTree(index);
 
-            long count = -1;
-            using var it = fst.Iterate();
+            // Dictionary training can be extremely IO intensive, therefore as illustrated in RavenDB-21106 we need to have
+            // an heuristic that allows us to get good dictionaries but at the same time minimizing the IO usage specially
+            // in IO deprived setups.
+            // https://issues.hibernatingrhinos.com/issue/RavenDB-21106
+
+            // If we page fault on every page (specially in cold scenarios) we would be wasting a lot. Therefore we should
+            // iterate with prefetching enabled.
+            using var it = fst.Iterate(prefetch: true);
             if (it.Seek(seek) == false)
                 yield break;
 
-            var result = new TableValueHolder();
-            while (it.MoveNext())
+            // Considering that in big collections the max amount of documents has been defined to be in the order of 100K 
+            // If we have to big skips we should get as many documents as possible to obtain as high locality as data allows.
+            int ChunkSize = 1024 * 16;
+
+            int ReadChunked(FixedSizeTree<long>.IFixedSizeIterator it, ByteString chunk, out bool hasMore)
             {
-                count++;
-                if (count % skip == 0)
+                // We will fill the buffer of the chunk expecting to have many documents in the chunk that end up 
+                // being stored in segments that are close to each other.
+                var chunkSpan = new Span<long>(chunk.Ptr, ChunkSize);
+                var keySpan = new Span<long>(chunk.Ptr + ChunkSize * sizeof(long), ChunkSize);
+                Debug.Assert(chunkSpan.Length == ChunkSize);
+                Debug.Assert(keySpan.Length == ChunkSize);
+
+                int readDocuments = 0;
+
+                do
                 {
-                    GetTableValueReader(it, out result.Reader);
-                    yield return (it.CurrentKey, result);
+                    keySpan[readDocuments] = it.CurrentKey;
+                    chunkSpan[readDocuments] = *(long*)it.ValuePtr(out int _);
+
+                    readDocuments++;
+
+                    hasMore = it.MoveNext();
                 }
+                while (readDocuments < chunkSpan.Length && hasMore);
+
+                if (readDocuments == 0)
+                    return 0;
+
+                Debug.Assert(readDocuments <= ChunkSize, "We cannot have more than ChunkSize because that is a buffer overflow.");
+
+                // We will divide the whole set into chunks of at least 128 skips (could be bigger if collection is small).
+                // Less than that and the tradeoff will be bad for IO (which we are trying to limit as much as possible). 
+                int localChunkSize = Math.Max(128, ChunkSize / ((int)skip + 1));
+                if (readDocuments < localChunkSize)
+                    return readDocuments;
```

Comment:
This should not be sorted to?

Reply:
I believe this was already answered here:
https://github.com/ravendb/ravendb/commit/d4456d6975d06e3cba443ccfccec9a9c00d0ff46#r125409585